### ISSUE TITLE
Increase batch size results in a non-linear increase in computation time

### ISSUE THREAD

**Issue by Wonder1905** [QwenLM/Qwen2-VL#616]
HI, I noticed that increasing the batch size train or inf results in a non-linear increase in computation time (we would except that linear increase would be an upper bound in some sense).
Ive saw it on my own environment, built another one and in the end tried also in the colab, here is the colab code:
```
from transformers import Qwen2VLForConditionalGeneration, AutoTokenizer, AutoProcessor
from qwen_vl_utils import process_vision_info
import numpy as np
import time
import torch
model = Qwen2VLForConditionalGeneration.from_pretrained(
    "Qwen/Qwen2-VL-2B-Instruct", torch_dtype="float16", device_map="auto"
)

processor = AutoProcessor.from_pretrained("Qwen/Qwen2-VL-2B-Instruct", torch_dtype="float16")
messages1 = [
    {
        "role": "user",
        "content": [
            {"type": "image", "resized_height": 256, "resized_width": 256, "image": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg"},
            {"type": "image", "resized_height": 256, "resized_width": 256, "image": "https://qianwen-res.oss-cn-beijing.aliyuncs.com/Qwen-VL/assets/demo.jpeg"},
            {"type": "text", "text": "What are the common elements in these pictures?"},
        ],
    }
]


#messages = [messages1,messages1,messages1,messages1]
#messages = [messages1,messages1,messages1]
#messages = [messages1,messages1]
messages = [messages1]
# Preparation for batch inference
texts = [
    processor.apply_chat_template(msg, tokenize=False, add_generation_prompt=True)
    for msg in messages
]
image_inputs, video_inputs = process_vision_info(messages)
inputs = processor(
    text=texts,
    images=image_inputs,
    videos=video_inputs,
    padding=True,
    return_tensors="pt",
)
inputs = inputs.to("cuda")

avg_list = []
for attempt in range(100):
    with torch.no_grad():
        torch.cuda.synchronize()
        start = time.perf_counter()  

        generated_ids = model(**inputs)
        torch.cuda.synchronize()
        end = time.perf_counter()
        avg_list.append(end-start)
        print(f"dotproduct time: {end - start}")
print("Avg of 100:",np.mean(avg_list))
```

Since colab can be tricky in allocating his resources, I did 10 run of 100  iterations and removed outliers the results were:
Batch=4: 1.4s
Batch=3: 0.88s
Batch=2: 0.51s
Batch=1: 0.19s
We can see the non linear increase, further when the size of the image increase it is much sharper.
After debugging, it happens in the visual:
self.visual = Qwen2VisionTransformerPretrainedModel._from_config(config.vision_config)

IN the scaled_dot_product:
attn_output = F.scaled_dot_product_attention(q, k, v, attention_mask, dropout_p=0.0)

And probably becuase you are treating the batch of images as a sequence and try to handle it with attention mask, but the seq length is biggest pain in transformers why is this the implementation?

Am I missing something?

**Comment by zwplus** [comment]
I encountered the same issue and noticed that the VisionTransformer in Qwen2-VL treats batched images as a single sequence. This leads to significant increases in memory usage during training, especially with larger batch sizes and high-resolution images. To address this, I modified the VisionTransformer in Qwen2-VL to use batch-based processing logic. Everything seems to work normally now. I also tested it using your script, and the inference time looks much more reasonable. I hope this solution helps!

**Comment by Wonder1905** [comment]
Any chance sharing the modified code?

**Comment by zwplus** [comment]
First of all, I apologize for the delayed response. I made some modifications to the modeling_qwen2_vl.py file in the transformers library to address this issue. Below are the changes I made.
### 1. Modified forward Function in Qwen2VisionTransformerPretrainedModel
Here is the updated code:
```python
class Qwen2VisionTransformerPretrainedModel(Qwen2VLPreTrainedModel):
    ...
    ...
    def forward(self, hidden_states: torch.Tensor, grid_thw: torch.Tensor) -> torch.Tensor:
        hidden_states = self.patch_embed(hidden_states)
        rotary_pos_emb = self.rot_pos_emb(grid_thw)

        cu_seqlens = torch.repeat_interleave(grid_thw[:, 1] * grid_thw[:, 2], grid_thw[:, 0]).cumsum(
            dim=0, dtype=torch.int32
        )
        cu_seqlens = F.pad(cu_seqlens, (1, 0), value=0)
        
        # Modified: Seq===>Batch
        singel_image_seq_length = int(grid_thw[0, 1] * grid_thw[0, 2])
        seq_length_sum, _ = hidden_states.shape
        batch_size = seq_length_sum // singel_image_seq_length
        hidden_states = hidden_states.reshape(batch_size, singel_image_seq_length, -1)  # B, h*w, embed_dim
        rotary_pos_emb = rotary_pos_emb.reshape(batch_size, singel_image_seq_length, -1)  # B, h*w, head_dim

        for blk in self.blocks:
            hidden_states = blk(hidden_states, cu_seqlens=cu_seqlens, rotary_pos_emb=rotary_pos_emb)
        
        # Modified: Batch===>Seq
        hidden_states = hidden_states.reshape(batch_size * singel_image_seq_length, -1)
        return self.merger(hidden_states)
```
### 2. Modified forward Function in VisionSdpaAttention
Here is the updated code:
```python
class VisionSdpaAttention(nn.Module):
    ...
    ...
    def forward(
        self, hidden_states: torch.Tensor, cu_seqlens: torch.Tensor, rotary_pos_emb: torch.Tensor = None
    ) -> torch.Tensor:
        batch_size, seq_length, _ = hidden_states.shape
        
        # Modified to support Batch processing
        q, k, v = self.qkv(hidden_states).reshape(batch_size, seq_length, 3, self.num_heads, -1).permute(0, 2, 3, 1, 4).unbind(1)
        q = apply_rotary_pos_emb_vision(q, rotary_pos_emb)
        k = apply_rotary_pos_emb_vision(k, rotary_pos_emb)
        
        # Adjusted attention mask
        attention_mask = torch.ones([batch_size, 1, seq_length, seq_length], device=q.device, dtype=torch.bool)
        attn_output = F.scaled_dot_product_attention(q, k, v, attention_mask, dropout_p=0.0)
        attn_output = attn_output.transpose(1, 2).reshape(batch_size, seq_length, -1)
        attn_output = self.proj(attn_output)
        return attn_output
```
### 3. Modified apply_rotary_pos_emb_vision Function
Here is the updated code:
```python
def apply_rotary_pos_emb_vision(tensor: torch.Tensor, freqs: torch.Tensor) -> torch.Tensor:
    orig_dtype = tensor.dtype
    tensor = tensor.float()  # B, head, seq, head_dim
    cos = freqs.cos()  # B, seq, head_dim/2
    sin = freqs.sin()  # B, seq, head_dim/2

    cos = cos.unsqueeze(1).repeat(1, 1, 1, 2).float()  # B, 1, seq, head_dim
    sin = sin.unsqueeze(1).repeat(1, 1, 1, 2).float()  # B, 1, seq, head_dim

    output = (tensor * cos) + (rotate_half(tensor) * sin)
    return output.to(orig_dtype)  # B, seq, head, head_dim
```
### Notes

  **1. This modification requires that all images in a batch have the same size.
  2. If you need to use FlashAttention, further changes may be necessary.**

I’ve done some simple tests, and these modifications seem to work. I hope this helps!

**Comment by liuyifan22** [comment]
Hello! We implemented a batched version for Qwen2.5-VL, which support batched image encoding. Hope it can help!

https://github.com/liuyifan22/Qwen2.5-VL-Batched
